internal/attestations/authorization.go

```diff
@@ -7,67 +7,47 @@ import (
-// attestation.
-type ReferenceAuthorization struct {
-	TargetRef      string `json:"targetRef"`
-	FromRevisionID string `json:"fromRevisionID"`
-	TargetTreeID   string `json:"targetTreeID"`
+// NewReferenceAuthorizationForCommit creates a new reference authorization for
+// the provided information. The authorization is embedded in an in-toto
+// "statement" and returned with the appropriate "predicate type" set. The
+// `fromID` and `targetID` specify the change to `targetRef` that is to be
+// authorized by invoking this function. The targetID is expected to be the Git
+// tree ID of the resultant commit.
+func NewReferenceAuthorizationForCommit(targetRef, fromID, targetID string) (*ita.Statement, error) {
+	return authorizations.NewReferenceAuthorizationForCommit(targetRef, fromID, targetID)
 }
 
-// NewReferenceAuthorization creates a new reference authorization for the
+// NewReferenceAuthorizationForTag creates a new reference authorization for the
 // provided information. The authorization is embedded in an in-toto "statement"
-// and returned with the appropriate "predicate type" set. The `fromTargetID`
-// and `toTargetID` specify the change to `targetRef` that is to be authorized
-// by invoking this function.
-func NewReferenceAuthorization(targetRef, fromRevisionID, targetTreeID string) (*ita.Statement, error) {
-	predicate := &ReferenceAuthorization{
-		TargetRef:      targetRef,
-		FromRevisionID: fromRevisionID,
-		TargetTreeID:   targetTreeID,
-	}
-
-	predicateStruct, err := predicateToPBStruct(predicate)
-	if err != nil {
-		return nil, err
-	}
-
-	return &ita.Statement{
-		Type: ita.StatementTypeUri,
-		Subject: []*ita.ResourceDescriptor{
-			{
-				Digest: map[string]string{digestGitTreeKey: targetTreeID},
-			},
-		},
-		PredicateType: ReferenceAuthorizationPredicateType,
-		Predicate:     predicateStruct,
-	}, nil
+// and returned with the appropriate "predicate type" set. The `fromID` and
+// `targetID` specify the change to `targetRef` that is to be authorized by
+// invoking this function. The targetID is expected to be the ID of the commit
+// the tag will point to.
+func NewReferenceAuthorizationForTag(targetRef, fromID, targetID string) (*ita.Statement, error) {
+	return authorizations.NewReferenceAuthorizationForTag(targetRef, fromID, targetID)
 }
 
 // SetReferenceAuthorization writes the new reference authorization attestation
 // to the object store and tracks it in the current attestations state.
-func (a *Attestations) SetReferenceAuthorization(repo *gitinterface.Repository, env *sslibdsse.Envelope, refName, fromRevisionID, targetTreeID string) error {
-	if err := validateReferenceAuthorization(env, refName, fromRevisionID, targetTreeID); err != nil {
+func (a *Attestations) SetReferenceAuthorization(repo *gitinterface.Repository, env *sslibdsse.Envelope, refName, fromID, targetID string) error {
+	// We assume that since we're setting a new authorization, it's the latest
+	// version
```

Comment:
I actually found such a case when two users have distinct versions of the client. If you approve using older version with v0.1 and I want to add my signature to the existing attestation but I have a newer client with v0.2, it'll fail because the payload is still the older version. We need to handle this somehow. Ultimately, I'm not sure if we're worried about predicate version rollbacks, it may just be a schema thing, so we validate using whatever the claimed version is. I need to think about this.